go-app-ussd_nurse.js

```diff
@@ -134,6 +220,197 @@ go.app = function() {
         });
 
+    // REGISTRATION STATES
+
+        self.add('state_subscribe_self', function(name) {
+            return new ChoiceState(name, {
+                question: $("To register we need to collect, store & use your info. You may also get messages on public holidays & weekends. Do you consent?"),
+                choices: [
+                    new Choice('state_check_optout_reg', $('Yes')),
+                    new Choice('state_permission_denied', $('No')),
+                ],
+                next: function(choice) {
+                    self.im.user.set_answer("registrant", self.im.user.answers.operator);
+                    return choice.value;
+                }
+            });
+        });
+
+        self.add('state_subscribe_other', function(name) {
+            return new ChoiceState(name, {
+                question: $("We need to collect, store & use your friend's info. She may get messages on public holidays & weekends. Does she consent?"),
+                choices: [
+                    new Choice('state_msisdn', $('Yes')),
+                    new Choice('state_permission_denied', $('No')),
+                ],
+                next: function(choice) {
+                    return choice.value;
+                }
+            });
+        });
+
+        self.add('state_permission_denied', function(name) {
+            return new ChoiceState(name, {
+                question: $("You have chosen not to receive NurseConnect SMSs on this number and so cannot complete registration."),
+                choices: [
+                    new Choice('state_route', $('Main Menu'))
+                ],
+                next: function(choice) {
+                    return choice.value;
+                }
+            });
+        });
+
+        self.add('state_msisdn', function(name) {
+            var error = $("Sorry, the format of the mobile number is not correct. Please enter the mobile number again, e.g. 0726252020");
+            var question = $("Please enter the number you would like to register, e.g. 0726252020:");
+            return new FreeText(name, {
+                question: question,
+                check: function(content) {
+                    if (!utils.is_valid_msisdn(content, 0, 10)) {
+                        return error;
+                    }
+                },
+                next: function(content) {
+                    msisdn = utils.normalize_msisdn(content, '27');
+
+                    return is
+                    .get_or_create_identity({"msisdn": msisdn})
+                    .then(function(identity) {
+                        self.im.user.set_answer("registrant", identity);
+                        return self.states.create('state_check_optout_reg');
+                    });
+                }
+            });
+        });
+
+        self.add('state_check_optout_reg', function(name) {
+            var registrant_msisdn = Object.keys(self.im.user.answers.registrant.details.addresses.msisdn)[0];
```

Comment:
Not sure this is safe using `[0]` - if the user has more than one msisdn (they could), this will fail.